frontend/src/utils/shootEditorCompletions.js

```diff
@@ -0,0 +1,293 @@
+          descWrapper.className = 'ghint-desc'
+          el.appendChild(descWrapper)
+        }
+      })
+    })
+    if (trim(token.string).length > 0) {
+      completionArray = filter(completionArray, completion => {
+        if (exactMatch) {
+          return isEqual(completion.string, token.string)
+        }
+        return includes(completion.string, token.string)
+      })
+    }
+    return completionArray
+  }
+
+  // get token at cursor position in editor with yaml content
+  _getYamlToken (cm, cur) {
+    let token
+    const { lineTokens, lineString } = this._getTokenLine(cm, cur)
+    forEach(lineTokens, lineToken => {
+      let result = lineToken.string.match(/^(\s*)-\s(.*)?$/)
+      if (result) {
+        const indent = result[1]
+        token = lineToken
+        token.type = 'firstArrayItem'
+        token.string = this._getTokenStringFromLine(lineString, cur.ch)
+        token.indent = token.start = indent.length
+        token.end = token.start + token.string.length + this.arrayBulletIndent
+        token.propertyName = trim(token.string)
+
+        const objectToken = this._returnObjectTokenFromTokenLine(lineTokens)
+        if (objectToken) {
+          // firstArrayItem line can also start new object
+          token.propertyName = objectToken.propertyName
+        }
+      }
+    })
+    if (!token) {
+      token = this._returnObjectTokenFromTokenLine(lineTokens)
+    }
+    if (!token) {
+      token = cm.getTokenAt(cur, true)
+      token.string = this._getTokenStringFromLine(lineString, cur.ch)
+      token.start = token.end - token.string.length
+    }
+    return token
+  }
+
+  // returns path to completions in shootCompletions object
+  _getTokenCompletionPath (token, cur, cm) {
+    let currentToken = token
+    let line = cur.line
+    let tokenContext = []
+
+    while (line >= 0 && !this._isTopLevelProperty(currentToken)) {
+      currentToken = this._getYamlToken(cm, CodeMirror.Pos(line, 0))
+      if (this._isCurrentTokenParentOfToken(currentToken, token) &&
+        this._isCurrentTokenIndentSmallerThanContextRoot(currentToken, tokenContext)) {
+        if (currentToken.type === 'property') {
+          tokenContext.unshift(currentToken)
+        } else if (currentToken.type === 'firstArrayItem') {
+          tokenContext.unshift(currentToken)
+        }
+      }
+      line--
+    }
+
+    // token context contains all parrent tokens of `token, except for the token itself
+    const tokenPath = flatMap(tokenContext, (pathToken, index, tokenContext) => {
+      const isLeafContextToken = pathToken === last(tokenContext)
+      if (pathToken.type === 'property') {
+        const nextToken = nth(tokenContext, index + 1)
+        if (nextToken && nextToken.type === 'firstArrayItem') {
+          // next item is array, so don't append 'properties' to path
+          return pathToken.propertyName
+        }
+        if (isLeafContextToken && token.type === 'firstArrayItem') {
```

Comment:
token or nextToken or currentToken? I am not sure?

Reply:
token is correct

Reply:
Do you want to go over this in more detail?